powerauth-restful-security-spring-annotation/src/main/java/io/getlime/security/powerauth/rest/api/spring/provider/CustomActivationProvider.java

```diff
@@ -108,15 +108,35 @@ default boolean shouldAutoCommitActivation(Map<String, String> identityAttribute
     default void activationWasCommitted(Map<String, String> identityAttributes, Map<String, Object> customAttributes, String activationId, String userId, String appId, ActivationType activationType, Map<String, Object> context) throws PowerAuthActivationException {}
 
     /**
-     * Method that indicates if the recovery codes should be revoked when an activation is removed.
+     * Method that indicates if recovery codes should be generated for a given activation or not. The method may return null,
+     * in such case, it uses settings of the PowerAuth Server to determine if the recovery codes should be generated or not. Also,
+     * just specifying true in the call will not result in generating recovery codes in case that recovery codes are
+     * globally disabled at the PowerAuth Server.
+     *
+     * @param identityAttributes Identity related attributes.
+     * @param customAttributes Custom attributes, not related to identity.
+     * @param activationType Activation type.
+     * @param context Context for passing parameters between activation provider calls.
+     * @return False to prevent generating recovery codes, "null" to let the PowerAuth Server decide, and true to generate recovery codes
+     *         in case that the feature is enabled globally on PowerAuth Server.
+     * @throws PowerAuthActivationException In case of error in custom activation business logic that should terminate the rest of the activation.
+     */
+    default Boolean shouldCreateRecoveryCodes(Map<String, String> identityAttributes, Map<String, Object> customAttributes, ActivationType activationType, Map<String, Object> context) throws PowerAuthActivationException {
```

Comment:
Just a philosophical question. I do not like three-state logic and taking care of NPE. Have you consider to return for example enum?

Reply:
I spent ~hour cracking this myself. The point is that in case someone decides to override the method, they typically want to return true / false.